mlir/lib/Dialect/Affine/Transforms/FoldMemRefAliasOps.cpp

```diff
@@ -0,0 +1,267 @@
+/// underlying indices (expanding the affine map).
+static void expandToUnderlyingIndices(AffineMap affineMap, ValueRange indices,
+                                      Location loc, PatternRewriter &rewriter,
+                                      SmallVectorImpl<Value> &result) {
+  SmallVector<OpFoldResult> indicesOfr(
+      llvm::map_to_vector(indices, [](Value v) -> OpFoldResult { return v; }));
+  for (unsigned i = 0, e = affineMap.getNumResults(); i < e; ++i) {
+    OpFoldResult ofr = affine::makeComposedFoldedAffineApply(
+        rewriter, loc, affineMap.getSubMap({i}), indicesOfr);
+    result.push_back(getValueOrCreateConstantIndexOp(rewriter, loc, ofr));
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// Patterns
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+class AffineLoadOpOfSubViewOpFolder final
+    : public OpRewritePattern<AffineLoadOp> {
+public:
+  using OpRewritePattern<AffineLoadOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(AffineLoadOp loadOp,
+                                PatternRewriter &rewriter) const override {
+    auto subViewOp = loadOp.getMemref().getDefiningOp<memref::SubViewOp>();
+
+    if (!subViewOp)
+      return rewriter.notifyMatchFailure(loadOp, "not a subview producer");
+
+    SmallVector<Value> indices;
+    expandToUnderlyingIndices(loadOp.getAffineMap(), loadOp.getIndices(),
+                              loadOp.getLoc(), rewriter, indices);
+
+    SmallVector<Value> sourceIndices;
+    affine::resolveIndicesIntoOpWithOffsetsAndStrides(
+        rewriter, loadOp.getLoc(), subViewOp.getMixedOffsets(),
+        subViewOp.getMixedStrides(), subViewOp.getDroppedDims(), indices,
+        sourceIndices);
+
+    rewriter.replaceOpWithNewOp<AffineLoadOp>(loadOp, subViewOp.getSource(),
+                                              sourceIndices);
+    return success();
+  }
+};
+
+class AffineLoadOpOfExpandShapeOpFolder final
+    : public OpRewritePattern<AffineLoadOp> {
+public:
+  using OpRewritePattern<AffineLoadOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(AffineLoadOp loadOp,
+                                PatternRewriter &rewriter) const override {
+    auto expandShapeOp =
+        loadOp.getMemref().getDefiningOp<memref::ExpandShapeOp>();
+
+    if (!expandShapeOp)
+      return failure();
+
+    SmallVector<Value> indices;
+    expandToUnderlyingIndices(loadOp.getAffineMap(), loadOp.getIndices(),
+                              loadOp.getLoc(), rewriter, indices);
+
+    SmallVector<Value> sourceIndices;
+    // affine.load guarantees that indexes start inbounds, which impacts if our
+    // linearization is `disjoint`.
+    if (failed(memref::resolveSourceIndicesExpandShape(
+            loadOp.getLoc(), rewriter, expandShapeOp, indices, sourceIndices,
+            /*startsInbounds=*/true)))
+      return failure();
```

Comment:
This is technically incorrect as `expandToUnderlyingIndices` modifies IR and we cannot return `failure()` anymore. But `resolveSourceIndicesExpandShape`/`resolveSourceIndicesCollapseShape` currently can never return `failure()` so maybe just mark them `void`?

Reply:
Done